### PR TITLE
In Pillar, add new Makefile targets for Docker dev builds and shell access.

### PR DESCRIPTION
**Pull request by OhmSpectator**
The Makefile has been updated to include two new targets: `build-docker-dev` and `enter-docker-dev`. These targets are designed to facilitate the evaluation of the codebase's state during the build process.

The `build-docker-dev` target constructs a Docker image up to the 'build' stage, enabling developers to assess the intermediate state of the code, particularly useful when applying patches that might only be present in temporary containers and are difficult to inspect otherwise.

Following the build, the `enter-docker-dev` target allows for immediate shell access within this intermediate Docker container, streamlining the process for developers to interact with the code and build environment directly.

### CODE REVIEW

**Comment by rouming** [comment]
I delegate the review to @christoph-zededa. He is the docker-pocker expert.

File: pkg/pillar/Makefile
Comment (christoph-zededa) [inline pkg/pillar/Makefile:83]: I normally add:
```
 -v $(pwd):$(pwd) 
```
this way I don't have to rebuild the container if I changed a file

**Review by christoph-zededa — APPROVED** [review]

**Comment by christoph-zededa** [comment]
> I delegate the review to @christoph-zededa. He is the docker-pocker expert.

LGTM

File: pkg/pillar/Makefile
Comment (OhmSpectator) [reply pkg/pillar/Makefile:83]: And how do you use it then? As far as I understand it will mount the host dir /pkg/pillar to / in the container, right?

Comment (christoph-zededa) [reply pkg/pillar/Makefile:83]: it mounts it to the same path in the container, f.e. for me it is: `/home/christoph/projects/eve/pkg/pillar`; then I do:
``` 
cd /path/to/mapped/directory
go test ./...
```

Comment (OhmSpectator) [reply pkg/pillar/Makefile:83]: Hm... In this case, will the host files in ${EVE_ROOT}/pkg/pillar replaced by the ones in the container?
If that's the case, I'm concerned that IDEs detecting disk changes might go haywire trying to reindex the directory structure that has been modified. Do you have any suggestions on how we could make this process more convenient for the developers?

Comment (christoph-zededa) [reply pkg/pillar/Makefile:83]: nothing will be replaced, it is just a bind-mount

> Do you have any suggestions on how we could make this process more convenient for the developers?
Use a proper IDE? ;-)

File: pkg/pillar/Makefile
Comment (rene) [inline pkg/pillar/Makefile:80]: @OhmSpectator , I liked these targets, but unfortunately, they don't consider architecture, as the whole Makefile. In a nutshell, I cannot run enter-docker-dev for ARM.... You don't need to add it in this PR, but it would be good to have architecture support in the future, maybe we could create an issue/ticket for that...

Comment (OhmSpectator) [reply pkg/pillar/Makefile:80]: Hmmm... Good point. And just a question of curiosity, how do you make ARCH-specific steps for building Pillar now? Is there anything build-time specific in general?

File: pkg/pillar/Makefile
Comment (christoph-zededa) [reply pkg/pillar/Makefile:83]: It will replace it, but I guess the normal container image has nothing in `/home/user/....`

> I'm not a fan of the "works for me" approach. I would prefer a solution suitable for more people...

Then we should perhaps start first with investigating which IDEs have that problem and where exactly they fail.

P.S.: just feel free to ignore this discussion - you already have the approval from me.

Comment (OhmSpectator) [reply pkg/pillar/Makefile:83]: Yeah, sure, I hope we can merge the current simple solution.
Let's just use this thread to discuss what could the better solution look like.

So you say you mount `${HOST_EVE_ROOT}/pkg/pillar` into `${HOST_EVE_ROOT}/pkg/pillar` inside the container, right? In this case, you have 2 copies of source code in the container: 1 - /pillar/ - preprocessed code with the patches applied, 2 - `${HOST_EVE_ROOT}`. Right? But in this case, if you change a file on your host - it will be visible only in the second path. So, the patched files are still invisible from the host and the tools/IDEs running there. Right?

File: pkg/pillar/Makefile
Comment (rene) [reply pkg/pillar/Makefile:80]: It's handled inside the Dockerfile to setup corresponding GO variables and cross-compilers (when needed): https://github.com/lf-edge/eve/blob/master/pkg/pillar/Dockerfile#L86

Comment (OhmSpectator) [reply pkg/pillar/Makefile:80]: So, the problem is that TARGETARCH is not set from Makefile, yeah?

File: pkg/pillar/Makefile
Comment (christoph-zededa) [reply pkg/pillar/Makefile:83]: I normally do it from the eve directory, not pkg/pillar, but the consequences are the same, that I have the code twice in the container.
Yes, it will be only visible in the second path.
Yes, the patched files are invisible from the host and therefore also invisible for the IDEs.

Comment (christoph-zededa) [reply pkg/pillar/Makefile:83]: > Yeah, sure, I hope we can merge the current simple solution.

Fine from my side.

File: pkg/pillar/Makefile
Comment (rene) [reply pkg/pillar/Makefile:80]: yes for `go build ...` commands, but in case of docker commands, it should be platform, like: `--platform linux/arm64`

Comment (eriknordmark) [reply pkg/pillar/Makefile:80]: I think this is the only remaining comment before this can be merged.

Comment (OhmSpectator) [reply pkg/pillar/Makefile:80]: Yeah. I tried to devise a quick fix, but it did not work as expected. I'll come up with another solution and push it here. Meanwhile, switching the PR into Draft.

Comment (OhmSpectator) [reply pkg/pillar/Makefile:80]: @eriknordmark, as the new targets are not the only ones that lack multi-arch support, I would prefer to address this issue in the scope of a dedicated task, not blocking the PR. I'll create a ticket for that shortly.

**Review by eriknordmark — APPROVED** [review]
LGTM but let's have @rene agree we can add arm64 support to the targets later.

**Pull request merged**
